In [1]:
import pandas as pd

DISPLAY_ALL_TEXT = False
pd.set_option("display.max_colwidth", 0 if DISPLAY_ALL_TEXT else 50)

In [2]:
with open('./my_amie_result.txt', 'r', encoding='utf-8') as file:
    lines = file.readlines()

with open('./my_amie_result.csv', 'w', encoding='utf-8') as file:
    for line in lines:
        line = line.replace('\t', ',')
        file.write(line)

In [3]:
rules_df = pd.read_csv('./my_amie_result.csv')
conf_sort_df = rules_df.sort_values('Std Confidence', ascending=False)
conf_sort_df

Rule  Head Coverage  \
17  ?a  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.466447   
5   ?g  <http://purl.org/spar/cito/cites>  ?a  ?g ...       0.149126   
6   ?a  <http://purl.org/spar/cito/cites>  ?h  ?h ...       0.171678   
15  ?g  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.466141   
0   ?b  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.989123   
12  ?g  <http://purl.org/dc/terms/creator>  ?a  ?g...       0.975143   
4   ?a  <http://purl.org/spar/cito/cites>  ?h  ?h ...       0.328358   
3   ?g  <http://purl.org/spar/cito/cites>  ?a  ?g ...       0.282416   
8   ?g  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.672168   
9   ?a  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.672485   
18  ?g  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.868423   
13  ?h  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.868745   
14  ?h  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.868584   
16  ?a  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.868745   
11  ?h  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.918270   
10  ?b  <http://lsdis.cs.uga.edu/projects/semdis/o...       0.918351   
2   ?h  <http://purl.org/spar/cito/cites>  ?b  ?a ...       0.196559   
1   ?a  <http://purl.org/spar/cito/cites>  ?h  ?h ...       0.143002   
7   ?g  <http://purl.org/spar/cito/cites>  ?a  ?g ...       0.120050   

    Std Confidence  PCA Confidence  Positive Examples  Body size  \
17        1.000000        1.000000               7632       7632   
5         1.000000        1.000000               2440       2440   
6         1.000000        1.000000               2809       2809   
15        1.000000        1.000000               7627       7627   
0         0.989123        0.989840              24552      24822   
12        0.744517        0.756123              24205      32511   
4         0.326531        0.381365               1408       4312   
3         0.309798        0.363336               1211       3909   
8         0.306114        0.407935               4236      13838   
9         0.305662        0.407461               4238      13865   
18        0.234088        0.234185              21556      92085   
13        0.233898        0.233949              21564      92194   
14        0.233855        0.233855              21560      92194   
16        0.233508        0.233508              21564      92348   
11        0.218118        0.219259              11314      51871   
10        0.217672        0.218812              11315      51982   
2         0.129012        0.129012               1451      11247   
1         0.121140        0.123117               2542      20984   
7         0.102670        0.104598               2134      20785   

    PCA Body size  Functional variable  
17           7632                   -1  
5            2440                   -1  
6            2809                   -1  
15           7627                   -1  
0           24804                   -1  
12          32012                   -1  
4            3692                   -1  
3            3333                   -1  
8           10384                   -1  
9           10401                   -1  
18          92047                   -1  
13          92174                   -1  
14          92194                   -1  
16          92348                   -1  
11          51601                   -2  
10          51711                   -2  
2           11247                   -1  
1           20647                   -1  
7           20402                   -1

In [4]:
for rule in conf_sort_df['Rule']:
    print(rule)

?a  <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor>  ?h  ?h  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b   => ?a  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b
?g  <http://purl.org/spar/cito/cites>  ?a  ?g  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b   => ?a  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b
?a  <http://purl.org/spar/cito/cites>  ?h  ?h  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b   => ?a  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b
?g  <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor>  ?a  ?g  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b   => ?a  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>  ?b
?b  <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor>  ?a   => ?a  <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor>  ?b
?g  <http://purl.org/dc/terms/creator>  ?a  ?g  <http://purl.org/dc/terms/creator>  ?b   => ?a  <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor>  ?b
?a  <http:/

# Chosen Rules - 1 per predicate:
---

### <u>**type** - $100\%$ <small>(high confidence)</small></u>

?a <http://purl.org/spar/cito/cites> ?h ?h <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?b => ?a <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?b \

```
INSERT {
    ?paperA a ?typeB .
} WHERE {
    ?paperA <http://purl.org/spar/cito/cites> ?paperH .
    ?paperH a ?typeB .
    FILTER NOT EXISTS{?paperA a ?typeB} .
}
```

<br>

---

<br>

### <u>**coauthor** - $98.9\%$ <small>(high confidence)</small></u>

?b <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?a => ?a <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?b

```
INSERT {
    ?authorA <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorB .
} WHERE {
    ?authorB <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorA .
    FILTER NOT EXISTS{?authorA <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorB} .
}
```
<br>

---

<br>

### <u>**appearsInConferenceSeries** - $32.7\%$ <small>(medium confidence)</small></u>

?a  <http://purl.org/spar/cito/cites> ?h ?h <https://makg.org/property/appearsInConferenceSeries> ?b => ?a <https://makg.org/property/appearsInConferenceSeries> ?b

```
INSERT {
    ?paperA <https://makg.org/property/appearsInConferenceSeries> ?seriesB .
} WHERE {
    ?paperA <http://purl.org/spar/cito/cites> ?paperH .
    ?paperH <https://makg.org/property/appearsInConferenceSeries> ?seriesB .
    FILTER NOT EXISTS{?paperA <https://makg.org/property/appearsInConferenceSeries> ?seriesB} .
}
```

<br>

---

<br>

### <u>**memberOf** - $30.6\%$ <small>(medium condfidence)</small></u>

?g  <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor>  ?a  ?g  <http://www.w3.org/ns/org#memberOf>  ?b   => ?a  <http://www.w3.org/ns/org#memberOf>  ?b

```
INSERT {
    ?authorA <http://www.w3.org/ns/org#memberOf> ?affiliationB .
} WHERE {
    ?authorG <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorA .
    ?authorG <http://www.w3.org/ns/org#memberOf> ?affiliationB .
    FILTER NOT EXISTS{?authorA <http://www.w3.org/ns/org#memberOf> ?affiliationB} .
}
```

<br>

---

<br>

### <u>**creator** - $21.8\%$ <small>(low condfidence)</small></u>

?h <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?b ?a <http://purl.org/dc/terms/creator> ?h => ?a <http://purl.org/dc/terms/creator> ?b

```
INSERT {
    ?paperA <http://purl.org/dc/terms/creator> ?authorB .
} WHERE {
    ?authorH <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorB .
    ?paperA <http://purl.org/dc/terms/creator> ?authorH .
    FILTER NOT EXISTS{?paperA <http://purl.org/dc/terms/creator> ?authorB} .
}
```


---

# Running the Queries

In [5]:
typeQuery = """
INSERT {
    ?paperA a ?typeB .
} WHERE {
    ?paperA <http://purl.org/spar/cito/cites> ?paperH .
    ?paperH a ?typeB .
    FILTER NOT EXISTS{?paperA a ?typeB} .
}
"""

coauthorQuery = """
INSERT {
    ?authorA <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorB .
} WHERE {
    ?authorB <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorA .
    FILTER NOT EXISTS{?authorA <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorB} .
}
"""

appearsInConfereSeriesQuery = """
INSERT {
    ?paperA <https://makg.org/property/appearsInConferenceSeries> ?seriesB .
} WHERE {
    ?paperA <http://purl.org/spar/cito/cites> ?paperH .
    ?paperH <https://makg.org/property/appearsInConferenceSeries> ?seriesB .
    FILTER NOT EXISTS{?paperA <https://makg.org/property/appearsInConferenceSeries> ?seriesB} .
}
"""

memberOfQuery = """
INSERT {
    ?authorA <http://www.w3.org/ns/org#memberOf> ?affiliationB .
} WHERE {
    ?authorG <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorA .
    ?authorG <http://www.w3.org/ns/org#memberOf> ?affiliationB .
    FILTER NOT EXISTS{?authorA <http://www.w3.org/ns/org#memberOf> ?affiliationB} .
}
"""

creatorQuery = """
INSERT {
    ?paperA <http://purl.org/dc/terms/creator> ?authorB .
} WHERE {
    ?authorH <http://lsdis.cs.uga.edu/projects/semdis/opus#coauthor> ?authorB .
    ?paperA <http://purl.org/dc/terms/creator> ?authorH .
    FILTER NOT EXISTS{?paperA <http://purl.org/dc/terms/creator> ?authorB} .
}
"""

In [6]:
import copy
from rdflib import Graph

g = Graph()
g.parse('train.nt', format='nt')
print(f'Original Graph Triple Count: {len(g)}\n')
gnew = copy.deepcopy(g)

for q in [typeQuery, coauthorQuery, appearsInConfereSeriesQuery, memberOfQuery, creatorQuery]:
    len_counter = len(gnew)
    gnew.update(q)
    print(f'Inserted Triples: {len(gnew) - len_counter}')
    print(f'New Graph Triple Count: {len(gnew)}')
    print()

Original Graph Triple Count: 89253

Inserted Triples: 0
New Graph Triple Count: 89253

Inserted Triples: 270
New Graph Triple Count: 89523

Inserted Triples: 2904
New Graph Triple Count: 92427

Inserted Triples: 9716
New Graph Triple Count: 102143

Inserted Triples: 41105
New Graph Triple Count: 143248

